oak/server/oak_node.cc

```diff
@@ -282,55 +273,26 @@ ::wabt::interp::HostFunc::Callback OakNode::OakWrite(wabt::interp::Environment*
     for (auto const& arg : args) {
       LOG(INFO) << "Arg: " << wabt::interp::TypedValueToString(arg);
     }
+    uint32_t offset = args[0].get_i32();
+    uint32_t size = args[1].get_i32();
 
-    // TODO: Synchronise this method.
-
-    uint32_t p = args[0].get_i32();
-    uint32_t len = args[1].get_i32();
-
-    auto data = ReadMemory(env, p, len);
-    response_data_->insert(response_data_->end(), data.cbegin(), data.cend());
-
-    results[0].set_i32(len);
+    ::absl::Span<const char> memory = ReadMemory(env, offset, size);
+    module_data_output_->insert(module_data_output_->end(), memory.cbegin(), memory.cend());
+    results[0].set_i32(size);
 
     return wabt::interp::Result::Ok;
   };
 }
 
-// Converts a gRPC ByteBuffer into a vector of bytes.
-// TODO: Move to GrpcStream.
-static std::unique_ptr<std::vector<char>> Unwrap(const ::grpc::ByteBuffer* buffer) {
-  auto bytes = absl::make_unique<std::vector<char>>();
-  std::vector<::grpc::Slice> slices;
-  ::grpc::Status status = buffer->Dump(&slices);
-  if (!status.ok()) {
-    LOG(QFATAL) << "Could not unwrap buffer";
-  }
-  for (const auto& slice : slices) {
-    bytes->insert(bytes->end(), slice.begin(), slice.end());
-  }
-  return bytes;
-}
-
-// Converts a vector of bytes into a gRPC ByteBuffer.
-// TODO: Move to GrpcStream.
-static const ::grpc::ByteBuffer Wrap(const std::vector<char>* bytes) {
-  ::grpc::Slice slice(bytes->data(), bytes->size());
-  ::grpc::ByteBuffer buffer(&slice, /*nslices=*/1);
-  return buffer;
-}
-
-// TODO: Refactor Oak Module code into a separate class.
-void OakNode::ProcessModuleCall(::grpc::GenericServerContext* context, ::grpc::ByteBuffer* request,
-                                ::grpc::ByteBuffer* response) {
-  // TODO: Synchronise this method.
+::grpc::Status OakNode::ProcessModuleCall(::grpc::GenericServerContext* context,
+                                          const std::vector<uint8_t>& request_data,
+                                          std::vector<uint8_t>* response_data) {
   LOG(INFO) << "Handling gRPC call: " << context->method();
   server_context_ = context;
 
-  request_data_ = Unwrap(request);
-  request_data_cursor_ = 0;
-
-  response_data_ = absl::make_unique<std::vector<char>>();
+  ::absl::MutexLock lock(&module_data_mutex_);
```

Comment:
Despite my comments on synchronization, I think we actually don't need to synchronize anything yet (until we introduce concurrency in the Wasm execution, which is not happening any time soon). Basically everything is already serialized, because we execute the gRPC loop sequentially, AFAICT?

Reply:
Yes, that should be the case.  This would be guard against that changing inadvertently, but we can remove it if you prefer.

Reply:
Let's leave it in for now, but we should make up our mind whether we care or not about synchronization at some point :)